File: services/undx_self_knowledge.py
"""Server-authoritative answer to "what is UNDX, and what can it do right now?"

The native and web clients must not decide capability availability from local
hard-coded metadata (mission Section 6/24). This module composes that answer on
the server from two authoritative sources:

* ``services.undx_company_identity`` — who builds PulseSoc and the honesty rules.
* ``services.undx_capability_registry`` — the allowlist of executable actions.

The capability registry is an allowlist whose stated invariant is that a
capability whose executor is not finished is *not registered* (it surfaces as
``unsupported_capability`` instead). So every entry here is genuinely executable:
its status is ``AVAILABLE``. Capabilities that are only planned or being
integrated are, by construction, absent from the registry — which is exactly why
UNDX must never claim to perform them. This module reports what is real, and the
company grounding block instructs UNDX to be honest about everything that is not.
"""

from __future__ import annotations

from typing import Any

from services import undx_capability_lifecycle as lifecycle
from services import undx_capability_registry as registry
from services import undx_company_identity as company
from services.undx_agent_contracts import ConfirmationPolicy, RiskLevel


def _domain_of(capability_id: str) -> str:
    """Derive a stable domain from the dotted capability id (e.g. crypto.alerts.create)."""
    head = str(capability_id or "").split(".", 1)[0].strip()
    return head or "system"


def _capability_view(spec: registry.CapabilitySpec) -> dict[str, Any]:
    is_write = spec.is_write
    return {
        "capability_id": spec.capability_id,
        "description": spec.description,
        "domain": _domain_of(spec.capability_id),
        # Every registered capability is executable by the registry's own invariant.
        "status": "AVAILABLE",
        "executionMode": "EXECUTE" if is_write else "READ",
        "requiresConfirmation": spec.confirmation != ConfirmationPolicy.NEVER,
        # Writes are independently verified; reads have no verifier by design.
        "requiresVerification": is_write,
        "receiptRequired": is_write,
    }


def capability_inventory() -> list[dict[str, Any]]:
    """Every executable capability, in a bounded, client-safe shape."""
    return [
        _capability_view(spec)
        for spec in sorted(registry.REGISTRY.values(), key=lambda s: s.capability_id)
    ]


def _counts(views: list[dict[str, Any]]) -> dict[str, Any]:
    by_domain: dict[str, int] = {}
    read = write = confirm = 0
    for view in views:
        by_domain[view["domain"]] = by_domain.get(view["domain"], 0) + 1
        if view["executionMode"] == "READ":
            read += 1
        else:
            write += 1
        if view["requiresConfirmation"]:
            confirm += 1
    return {
        "total": len(views),
        "read_only": read,
        "write": write,
        "requires_confirmation": confirm,
        "by_domain": dict(sorted(by_domain.items())),
    }


def self_knowledge() -> dict[str, Any]:
    """The bootstrap self-description UNDX/clients can render without hard-coding.

    Server-authoritative. Safe to return to an authenticated client: it exposes
    capability *ids and shapes*, never executors, verifiers, permissions internals,
    secrets, or metrics.
    """
    views = capability_inventory()
    return {
        "assistant": {
            "name": "UNDX",
            "description": (
                "UNDX is PulseSoc's native intelligence companion. It helps you "
                "understand and navigate the ecosystem and can perform a growing "
                "set of governed, verified PulseSoc actions."
            ),
        },
        "company": company.facts(),
        "canonical": {
            "company_explanation": company.CANONICAL_COMPANY_EXPLANATION,
            "pulsesoc_definition": company.CANONICAL_PULSESOC_DEFINITION,
        },
        "capabilities": {
            "counts": _counts(views),
            "available": views,
            # Full lifecycle projection (AVAILABLE/LIMITED/TRAINING/PLANNED/
            # DISABLED) over registry + knowledge map + live server policy.
            # Server-authoritative; supersedes the flat "available" list for
            # clients that render capability state.
            "lifecycle": lifecycle.lifecycle_inventory(),
            "lifecycle_counts": lifecycle.lifecycle_counts(),
            "canonical_language": dict(lifecycle.CANONICAL_STATUS_LANGUAGE),
        },
        "honesty": {
            "never_fabricates": list(company.UNVERIFIABLE_WITHOUT_SOURCE),
            "capability_rule": (
                "UNDX only reports actions this registry can actually execute and "
                "verify. Anything not listed here is not executable yet, and UNDX "
                "will say so rather than claim it was done."
            ),
        },
        "version": {
            "company_identity": company.COMPANY_IDENTITY_VERSION,
        },
    }

File: mobile-native/src/api/undxSelfKnowledge.ts
import { pulseApi } from "./pulseApi";

/**
 * Server-authoritative UNDX self-knowledge.
 *
 * The native app must NOT hard-code who builds PulseSoc or which UNDX actions are
 * available — that is decided on the server (see
 * `services/undx_self_knowledge.py :: self_knowledge`) and delivered inside the
 * `GET /api/pulse-ai/conversation` bootstrap payload. This module gives native
 * surfaces a typed, read-only view of that payload plus pure selectors, so a
 * screen can render live capability status without inventing or caching its own
 * copy of the truth.
 *
 * Shapes here mirror the Python composer exactly. Everything is optional-tolerant:
 * an older backend that predates `self_knowledge` simply yields `null`, and
 * callers must treat "unknown" as "not available" rather than assuming a
 * capability works.
 */

export type UndxExecutionMode = "READ" | "EXECUTE";
export type UndxCapabilityStatus = "AVAILABLE";

export type UndxCapabilityView = {
  capability_id: string;
  description: string;
  domain: string;
  status: UndxCapabilityStatus;
  executionMode: UndxExecutionMode;
  requiresConfirmation: boolean;
  requiresVerification: boolean;
  receiptRequired: boolean;
};

export type UndxCapabilityCounts = {
  total: number;
  read_only: number;
  write: number;
  requires_confirmation: number;
  by_domain: Record<string, number>;
};

export type UndxCompanyFacts = {
  version: number;
  legal_name: string;
  primary_product: string;
  founder: { name: string; title: string };
  product_category: string[];
};

export type UndxSelfKnowledge = {
  assistant: { name: string; description: string };
  company: UndxCompanyFacts;
  canonical: { company_explanation: string; pulsesoc_definition: string };
  capabilities: {
    counts: UndxCapabilityCounts;
    available: UndxCapabilityView[];
  };
  honesty: { never_fabricates: string[]; capability_rule: string };
  version: { company_identity: number };
};

type ConversationBootstrap = {
  self_knowledge?: UndxSelfKnowledge | null;
};

/**
 * Fetch the UNDX conversation bootstrap and return only its self-knowledge block.
 * Returns `null` when the backend does not supply one — callers must degrade to a
 * "capabilities unknown" state, never to a fabricated capability list.
 */
export async function fetchUndxSelfKnowledge(): Promise<UndxSelfKnowledge | null> {
  const data = await pulseApi<ConversationBootstrap>("/api/pulse-ai/conversation");
  return data?.self_knowledge ?? null;
}

/** True only when the server lists this capability as genuinely executable. */
export function isCapabilityAvailable(
  knowledge: UndxSelfKnowledge | null,
  capabilityId: string
): boolean {
  if (!knowledge) return false;
  return knowledge.capabilities.available.some(
    (view) => view.capability_id === capabilityId && view.status === "AVAILABLE"
  );
}

/** Capability ids the server currently advertises as executable. */
export function availableCapabilityIds(knowledge: UndxSelfKnowledge | null): string[] {
  if (!knowledge) return [];
  return knowledge.capabilities.available.map((view) => view.capability_id);
}

/** Group the advertised capabilities by their dotted-id domain, sorted by id. */
export function capabilitiesByDomain(
  knowledge: UndxSelfKnowledge | null
): Record<string, UndxCapabilityView[]> {
  const grouped: Record<string, UndxCapabilityView[]> = {};
  if (!knowledge) return grouped;
  for (const view of knowledge.capabilities.available) {
    (grouped[view.domain] ??= []).push(view);
  }
  for (const domain of Object.keys(grouped)) {
    grouped[domain].sort((a, b) => a.capability_id.localeCompare(b.capability_id));
  }
  return grouped;
}

/** Capabilities that must be confirmed before UNDX executes them. */
export function capabilitiesRequiringConfirmation(
  knowledge: UndxSelfKnowledge | null
): UndxCapabilityView[] {
  if (!knowledge) return [];
  return knowledge.capabilities.available.filter((view) => view.requiresConfirmation);
}

/**
 * A capability the model might name but that the server does NOT list is, by the
 * registry's allowlist invariant, not executable yet. Native surfaces should use
 * this to keep UNDX honest instead of trusting prose.
 */
export function isCapabilityClaimHonest(
  knowledge: UndxSelfKnowledge | null,
  capabilityId: string
): boolean {
  return isCapabilityAvailable(knowledge, capabilityId);
}
